service/src/PlatformTopo.cpp

```diff
@@ -463,28 +464,56 @@ namespace geopm
     void PlatformTopoImp::create_cache(void)
     {
         try {
-            PlatformTopoImp::create_cache(M_SERVICE_CACHE_FILE_NAME);
+            // TODO: Handle return string?
+            (void) PlatformTopoImp::create_cache(M_SERVICE_CACHE_FILE_NAME);
         }
         catch (const geopm::Exception &ex) {
-            PlatformTopoImp::create_cache(M_CACHE_FILE_NAME);
+            // TODO: Handle return string?
+            (void) PlatformTopoImp::create_cache(M_CACHE_FILE_NAME);
         }
     }
 
-    void PlatformTopoImp::create_cache(const std::string &cache_file_name)
+    std::string PlatformTopoImp::create_cache(const std::string &cache_file_name)
     {
-        // If cache file is not present, or is too old, create it
+        std::string contents;
         bool is_file_ok = false;
-        try {
-            is_file_ok = check_file(cache_file_name);
+
+        // Open fd
+        // if err, goto create
+        // if no err, pass to check_file
+        // if no err, return contents, else create and return contents
+
+        int cache_fd = open(cache_file_name.c_str(), O_RDWR);
+        if (cache_fd < 0) {
+            if (errno != ENOENT) {
+                throw Exception("PlatformTopoImp: Could not open cache file " + cache_file_name,
+                                errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+            }
         }
-        catch (const geopm::Exception &ex) {
-            // sysinfo or stat failed; file does not exist (2) or permission denied (13)
-            if (ex.err_value() == EACCES) {
-                throw; // Permission was denied; Cannot create files at the desired path
+        else {
+            try {
+                mode_t expected_perms;
+                if (cache_file_name == M_SERVICE_CACHE_FILE_NAME) {
+                    expected_perms = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH; // 0o644
+                }
+                else {
+                    expected_perms = S_IRUSR | S_IWUSR; // 0o600
+                }
+                is_file_ok = check_file(cache_fd, expected_perms);
+            }
+            catch (const geopm::Exception &ex) {
+                (void) close(cache_fd);
+                // sysinfo or stat failed; file does not exist (2) or permission denied (13)
+                throw;
```

Comment:
Since it's assured that the file exists now (since it was opened above) all errors in check_file() should throw all the way up.  If the file does not exist, this else block is not evaluated.